File: JAVA/Arrays/Day_8.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.stream.Gatherer.Integrator;

public class Day_8 {

    // Q1. Leaders Array
    public static ArrayList<Integer>  Leaders_Array(int[] arr){
        int maxi = Integer.MIN_VALUE;
        ArrayList <Integer> ans = new ArrayList<>();
        for(int i = arr.length-1; i >= 0;i--){
            if(arr[i] > maxi){
                ans.add( arr[i]);
            }
            maxi = Math.max(maxi, arr[i]);
        }
        return ans;
    }

    // Q2. Longest Consecutive element
    public static int Longest_Consecutive_Opt(int [] arr){
        int n = arr.length;
        int Longest = 0;
        HashSet<Integer> newSet = new HashSet<>();
        if (n ==0){
            return 0;
        }
        for (int el: arr){
            newSet.add(el);
        }
        
        for (int it : newSet){
            if(!newSet.contains(it -1)){
                int cnt = 1;
                int x = it;
                while(newSet.contains(x+1)){
                    cnt++;
                    x++;
                }
                Longest = Math.max(cnt, Longest);
            }
        }

        return Longest;

    }

    public static void print(ArrayList<Integer> arr){
      for(int i = 0; i < arr.size(); i++){
            System.out.print(arr.get(i)+" ");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        int [] arr = {10,22,12,3,0,6};
        print(Leaders_Array(arr));

        int [] arr2 = {102,4,100,1,101,3,2,1};
        System.out.println(Longest_Consecutive_Opt(arr2));


    }
}

File: PYTHON/Arrays/Day_8.py
# Q1. Leaders Array
def Leaders_array_Opt(arr):
    j = len(arr) -1
    i = len(arr) - 1
    print(arr[i])
    ans = []
    while j >= 0:
        if  j == i:
            ans.append(arr[j])
        if arr[i] < arr[j]:
            ans.append(arr[j])
            i = j
        print(j)
        j-=1

    return ans

def Leaders_array_Opt2(arr):
    maxi = float('-inf')
    ans = []
    for i in range(len(arr)-1, -1, -1):
        if arr[i] > maxi:
            ans.append(arr[i])
        maxi = max(maxi,arr[i])
    return ans

# Q2. Longest Consecutive element
def Longest_consecutive_Opt(arr):
    n = len(arr)
    if (n == 0): return 0
    st = set(arr)
    length = 0

    for it in st:
        if it-1 not in st:
            cnt = 1
            x = it
            while x+1 in st:
                x += 1
                cnt +=1 
            length = max(length,cnt)

    return length


arr = [10,22,12,3,0,6]
print(Leaders_array_Opt2(arr))

arr2 = [102,4,100,1,101,3,2,1]
print(Longest_consecutive_Opt(arr2))